### PR TITLE
python-cryptography 41.0.2 (new formula)

### PR DESCRIPTION
**Pull request by p-linnane**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----
Python formulae that use the `cryptography` PyPI package require Rust as a build dep. Packaging `cryptography` into a formula will allow us to remove the Rust dep from a large number of formulae.

### CODE REVIEW

**Comment by p-linnane** [comment]
This doesn't build currently because it wants `setuptools-rust`. I wasn't able to have `cryptography` find it after building it as a resource though. Looking for additional help here.

cc @alex for any thoughts.

**Comment by alex** [comment]
As I understand it, `--no-build-isolation` has the effect of not installing build deps, so you'd have to manually install `setuptools-rust` before installing cryptography.

File: Formula/python-cryptography.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    system python3, "-m", "pip", "install", "--prefix=#{prefix}", "--no-deps", "--no-build-isolation", "."
+  end
+
+  test do
+    (testpath/"test.py").write <<~EOS
+      from cryptography.fernet import Fernet
+      key = Fernet.generate_key()
+      f = Fernet(key)
+      token = f.encrypt(b"homebrew")
+      f.decrypt(token)
```
Comment (alex) [inline Formula/python-cryptography.rb:34]:
```suggestion
      print(f.decrypt(token))
```

**Comment by alex** [comment]
For the `setuptools_rust` installation, is it installing it somewhere that'll be on the python path when cryptography is installed?

**Review by chenrui333 — APPROVED** [review]

File: Formula/python-cryptography.rb [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+  def install
+    site_packages = prefix/Language::Python.site_packages(python3)
+    ENV.append_path "PYTHONPATH", site_packages
+
+    resources.each do |r|
+      r.stage do
+        system python3, "-m", "pip", "install", "--prefix=#{prefix}", "--no-deps", "--no-build-isolation", "."
+      end
+    end
```
Comment (Bo98) [inline Formula/python-cryptography.rb:52]: Is `setuptools_rust` etc really needed at runtime?

Comment (alex) [reply Formula/python-cryptography.rb:52]: It's not.

Comment (p-linnane) [reply Formula/python-cryptography.rb:52]: This was the only way I could figure out to get it to build. Any help to improve this is appreciated.

Comment (Bo98) [reply Formula/python-cryptography.rb:52]: `buildpath` instead of `prefix`?

Comment (p-linnane) [reply Formula/python-cryptography.rb:52]: https://github.com/Homebrew/homebrew-core/pull/137378

File: Formula/python-cryptography.rb [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  depends_on "rust" => :build
+  depends_on "cffi"
+  depends_on "openssl@3"
+  depends_on "pycparser"
+  depends_on "python-typing-extensions"
+  depends_on "python@3.11"
+
+  on_linux do
+    depends_on "pkg-config" => :build
+  end
```
Comment (carlocab) [inline Formula/python-cryptography.rb:28]: You don't need to put this inside an `on_linux` block. It's a dependency of `rust`, and the `pkg-config` shim for macOS means that it gets used on macOS anyway.

**Pull request merged**